test/e2e/usersignup_test.go

```diff
@@ -515,3 +519,67 @@ func (s *userSignupIntegrationTest) TestSkipSpaceCreation() {
 
 	VerifyResourcesProvisionedForSignupWithoutSpace(s.T(), s.Awaitilities, userSignup, "base")
 }
+
+func (s *userSignupIntegrationTest) TestUserSignupMigration() {
+	// Manually create a UserSignup that has a name that is not the encoded username
+
+	md5hash := md5.New() // nolint:gosec
+	// Ignore the error, as this implementation cannot return one
+	_, _ = md5hash.Write([]byte("foo@bar.com"))
+	emailHash := hex.EncodeToString(md5hash.Sum(nil))
+
+	userSignup := &toolchainv1alpha1.UserSignup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "userid-we-migrate-from",
+			Namespace: s.Host().Namespace,
+			Labels: map[string]string{
+				toolchainv1alpha1.UserSignupUserEmailHashLabelKey: emailHash,
+			},
+			Annotations: map[string]string{toolchainv1alpha1.UserSignupUserEmailAnnotationKey: "foo@bar.com"},
+		},
+		Spec: toolchainv1alpha1.UserSignupSpec{
+			Userid:   "userid-we-migrate-from",
+			Username: "username-we-migrate-to",
+		},
+	}
+
+	states.SetApproved(userSignup, true)
+
+	require.NoError(s.T(), s.Awaitilities.Host().Client.Create(context.TODO(), userSignup))
+
+	// Let the UserSignup provision
+	userSignup, err := s.Awaitilities.Host().WaitForUserSignup(userSignup.Name,
+		wait.UntilUserSignupContainsConditions(ApprovedByAdmin()...))
+	require.NoError(s.T(), err)
+
+	// Deactivate
+	states.SetDeactivated(userSignup, true)
+	userSignup, err = s.Awaitilities.Host().UpdateUserSignup(userSignup.Name, func(us *toolchainv1alpha1.UserSignup) {
+		states.SetDeactivated(us, true)
+	})
+	require.NoError(s.T(), err)
+
+	// The UserSignup should be migrated, so it is expected that a new UserSignup will be created
+	migrated, err := s.Awaitilities.Host().WaitForUserSignup("username-we-migrate-to")
+	require.NoError(s.T(), err)
+
+	require.NoError(s.T(), s.Awaitilities.Host().WaitUntilUserSignupDeleted(userSignup.Name))
+
+	// Now try to reactivate the migrated UserSignup again
+	_, err = s.Awaitilities.Host().UpdateUserSignup(migrated.Name, func(us *toolchainv1alpha1.UserSignup) {
+		us.Labels[toolchainv1alpha1.StateLabelKey] = "approved"
+		states.SetApproved(us, true)
+		states.SetDeactivated(us, false)
+	})
+	require.NoError(s.T(), err)
+
+	// Confirm that the migrated UserSignup is provisioned ok
+	migrated, err = s.Awaitilities.Host().WithRetryOptions(wait.TimeoutOption(time.Second*10)).WaitForUserSignup(migrated.Name,
+		wait.UntilUserSignupHasConditions(ConditionSet(Default(), ApprovedByAdmin())...),
+		wait.UntilUserSignupHasStateLabel(toolchainv1alpha1.UserSignupStateLabelValueApproved))
+	require.NoError(s.T(), err)
+
+	// Confirm that the MasterUserRecord is created
+	_, err = s.Awaitilities.Host().WaitForMasterUserRecord(migrated.Status.CompliantUsername)
+	require.NoError(s.T(), err)
```

Comment:
you can make it simpler:
```suggestion
	VerifyResourcesProvisionedForSignup(s.T(), s.Awaitilities, migrated, "base")
```

Reply:
Fixed